pkg/querier/metadata_handler.go

```diff
@@ -46,17 +47,47 @@ type metadataErrorResult struct {
 // Mimir for a given tenant. It is kept and returned as a set.
 func NewMetadataHandler(m MetadataSupplier) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		limit := -1
+		if s := r.FormValue("limit"); s != "" {
+			var err error
+			if limit, err = strconv.Atoi(s); err != nil {
+				w.WriteHeader(http.StatusBadRequest)
+				util.WriteJSONResponse(w, metadataErrorResult{Status: statusError, Error: "limit must be a number"})
+				return
+			}
+		}
+		limitPerMetric := -1
+		if s := r.FormValue("limit_per_metric"); s != "" {
```

Comment:
Would it be possible to send the limit per metric to ingesters and store-gateways, so they can limit their response?

(Not sure if we should then apply the limit here as well for backwards compatibility if we're talking to an old ingester or store-gateway that doesn't recognise the limit.)

Reply:
Added tests and updated changelog, but not sure how to do this part. I tried but ended up at a point where I'll probably need to modify the proto files to add the new config struct. Alternatively I could pass on the configs directly as ints and a string instead, but a struct seems neater (and if we need to modify the options in future, we won't need to make as many changes). I didn't see where store gateways are affected, and as for the ingester, the earliest part I can apply the config is https://github.com/grafana/mimir/pull/5890/files#diff-07bc881ce238d2cc7b32972e95be91bc4796f48528c6c9330ba473d183e553c4R94 , not sure if it's worth it to make a lot of changes everywhere for that. But there might be a better way, please let me know!

Reply:
Could we extend `MetricsMetadataRequest` in [`ingester.proto`](https://github.com/grafana/mimir/blob/ed439865fc70799cb321678f8d3133ce57d7fee7/pkg/ingester/client/ingester.proto#L219) to pass along the parameters?

And I think I may have led you astray with the mention of store-gateways, looks like we only ever query ingesters for metadata, which makes sense given [the description in the Prometheus docs](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-metric-metadata).

Reply:
Thanks, it works now